### PR TITLE
Parsing table of contents

### PR DESCRIPTION
**Pull request by misoxxx**

### CODE REVIEW

File: src/parsers/parse_toc.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    for (auto lines_since_last_toc_entry = 0;
+         lines_since_last_toc_entry < max_length_of_toc && (current_position = s.find('\n', previous_position + 1)) != std::string::npos;
+         ++lines_since_last_toc_entry)
+    {
+        const auto line = s.substr(previous_position, current_position - previous_position);
+
+        std::smatch split_matches_double_line, split_matches_single_line;
+        std::regex_match(line, split_matches_double_line, toc_double_split);
+        std::regex_match(line, split_matches_single_line, toc_split);
+        if (split_matches_double_line.size() >= 7 && split_matches_double_line[2].length() > 3)
```
Comment (mdugacek) [inline src/parsers/parse_toc.cpp:58]: It would be nice to create named constants from 7 and 3.

File: src/parsers/parse_toc.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            {
+                section_r.id.pop_back();
+            }
+            section_r.name        = split_matches_double_line[5];
+            section_r.page_number = stoi(split_matches_double_line[6]);
+            table_of_contents.push_back(section_r);
+            lines_since_last_toc_entry = 0;
+        }
+
+        else if (split_matches_single_line.size() >= 4 && split_matches_single_line[2].length() > 3)
```
Comment (mdugacek) [inline src/parsers/parse_toc.cpp:81]: Again, create named constants from 4 and 3.

File: src/parsers/parse_toc.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+} // namespace
+
+
+namespace sparse::parsers
+{
+std::optional<common::table_of_contents_t> parse_toc(const std::string& whole_file) noexcept
+{
+    const auto possible_toc = find_possible_toc_beginnings(whole_file);
+
+    common::table_of_contents_t biggest_toc;
```
Comment (mdugacek) [inline src/parsers/parse_toc.cpp:113]: If you create this as `std::optional` you do not need the `if` at the end of the function.

**Review by mdugacek — CHANGES_REQUESTED** [review]

File: src/parsers/parse_toc.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            section.name        = split_matches_single_line[2];
+            section.page_number = stoi_fallback(split_matches_single_line[3]);
+            table_of_contents.push_back(section);
+            lines_since_last_toc_entry = 0;
+        }
+
+        previous_position = current_position;
+    }
+
+    std::sort(table_of_contents.begin(), table_of_contents.end(), &toc_sort);
```
Comment (ondrys123) [inline src/parsers/parse_toc.cpp:119]: `toc_sort` is called only once, therefore I think lambda is preferable.
Also, body of this function can be simplified using `std::tie`:
```
return std::tie(lhs.page_number, lhs.id) < std::tie(rhs.page_number, rhs.id);
```

File: src/parsers/parse_toc.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+
+    for (const auto i : possible_toc)
+    {
+        /*
+         * Usually there are few false positives - for example if text bibliography is in table of contents
+         * In the case there is more of them, we try to parse each one
+         * And take the one that contains most entries, as it is the one that is most likely to not be a false positive
+         */
+        const auto possible_toc = get_toc(whole_file, i);
```
Comment (ondrys123) [inline src/parsers/parse_toc.cpp:144]: `possible_toc` shadows variable declared at line 132, choose a different name

**Review by ondrys123 — CHANGES_REQUESTED** [review]
Looks good, just needs a few tweaks.

**Review by mdugacek — APPROVED** [review]

**Review by ondrys123 — APPROVED** [review]
Is good

**Pull request merged**